File: node-health-ai/src/services/ai.service.js
// services/ai.service.js
// Native Node.js AI services — Gemini Vision (primary) + OpenAI (fallback)
// Replaces the Python FastAPI microservice

import fs from 'fs';
import OpenAI from 'openai';
import Groq from 'groq-sdk';
import { GoogleGenerativeAI } from '@google/generative-ai';

const GEMINI_MODEL = 'gemini-2.0-flash';

const parseFoodAnalysisJson = (text) => {
    if (!text) return null;

    const cleanedText = String(text)
        .replace(/```json/gi, '')
        .replace(/```/g, '')
        .trim();

    const jsonMatch = cleanedText.match(/\{[\s\S]*\}/);
    if (!jsonMatch) return null;

    try {
        return JSON.parse(jsonMatch[0]);
    } catch (error) {
        console.warn('⚠️ Failed to parse food analysis JSON:', error.message);
        return null;
    }
};

const runGroqVisionAnalysis = async (foodPrompt, base64image, mimeType) => {
    if (!process.env.GROQ_API_KEY) return null;

    const groq = new Groq({ apiKey: process.env.GROQ_API_KEY });
    const response = await groq.chat.completions.create({
        model: 'llama-3.2-11b-vision-preview',
        messages: [{
            role: 'user',
            content: [
                { type: 'text', text: foodPrompt },
                { type: 'image_url', image_url: { url: `data:${mimeType};base64,${base64image}` } }
            ]
        }],
        max_tokens: 400,
        temperature: 0.2,
    });

    const text = response.choices?.[0]?.message?.content || '';
    const data = parseFoodAnalysisJson(text);
    if (!data) return null;

    return { success: true, data, source: 'groq' };
};

// ─────────────────────────────────────────────────────────────────────────────
// GEMINI VISION — Food Image Analysis (non-streaming)
// ─────────────────────────────────────────────────────────────────────────────
export const analyseFoodImage = async (imagePath) => {
    const FOOD_PROMPT = `Analyze this food image and return ONLY valid JSON with exactly this structure:
{
    "foodName": "name of the food dish",
    "calories": 0,
    "protein": 0,
    "carbohydrates": 0,
    "fat": 0,
    "fiber": 0,
    "sugar": 0,
    "healthyScore": 0,
    "description": "Brief nutritional description",
    "portionSize": "Estimated portion size"
}
Only return the JSON object, no markdown, no other text.`;

    const imageData = fs.readFileSync(imagePath);
    const base64image = imageData.toString('base64');
    const mimeType = imagePath.endsWith('.png') ? 'image/png'
                   : imagePath.endsWith('.webp') ? 'image/webp' : 'image/jpeg';

    // ✅ Use the existing Groq key for food image analysis first
    if (process.env.GROQ_API_KEY) {
        try {
            const result = await runGroqVisionAnalysis(FOOD_PROMPT, base64image, mimeType);
            if (result?.success) {
                console.log(`✅ Groq food analysis: ${result.data.foodName}`);
                return result;
            }
        } catch (groqErr) {
            console.warn('⚠️ Groq food analysis failed, trying Gemini:', groqErr.message);
        }
    }

    // ✅ Try Gemini Vision next
    if (process.env.GEMINI_API_KEY) {
        try {
            const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
            const model = genAI.getGenerativeModel({ model: GEMINI_MODEL });

            const maxRetries = 3;
            for (let attempt = 0; attempt < maxRetries; attempt++) {
                try {
                    const result = await model.generateContent([
                        FOOD_PROMPT,
                        { inlineData: { mimeType, data: base64image } }
                    ]);
                    const text = result.response.text();
                    const data = parseFoodAnalysisJson(text);
                    if (data) {
                        console.log(`✅ Gemini food analysis: ${data.foodName}`);
                        return { success: true, data, source: 'gemini' };
                    }
                } catch (err) {
                    const msg = err.message || '';
                    if ((msg.includes('429') || msg.includes('RESOURCE_EXHAUSTED')) && attempt < maxRetries - 1) {
                        const wait = 2000 * Math.pow(2, attempt);
                        console.warn(`⚠️ Gemini rate limit, retrying in ${wait}ms...`);
                        await new Promise(r => setTimeout(r, wait));
                        continue;
                    }
                    throw err;
                }
            }
        } catch (geminiErr) {
            console.warn('⚠️ Gemini food analysis failed, trying OpenAI:', geminiErr.message);
        }
    }

    // ✅ Fallback: OpenAI GPT-4o-mini Vision
    if (process.env.OPENAI_API_KEY) {
        try {
            const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
            const response = await openai.chat.completions.create({
                model: 'gpt-4o-mini',
                messages: [{
                    role: 'user',
                    content: [
                        { type: 'text', text: FOOD_PROMPT },
                        { type: 'image_url', image_url: { url: `data:${mimeType};base64,${base64image}` } }
                    ]
                }],
                max_tokens: 350,
            });
            const text = response.choices[0].message.content;
            const data = parseFoodAnalysisJson(text);
            if (data) {
                console.log(`✅ OpenAI food analysis: ${data.foodName}`);
                return { success: true, data, source: 'openai' };
            }
        } catch (openaiErr) {
            console.error('❌ OpenAI food fallback failed:', openaiErr.message);
        }
    }

    return { success: false, message: 'Food analysis failed. Check API keys.' };
};

// ─────────────────────────────────────────────────────────────────────────────
// GEMINI VISION — Food Image Analysis (SSE streaming)
// ─────────────────────────────────────────────────────────────────────────────
export const analyseFoodImageStream = async (imagePath, res) => {
    const FOOD_PROMPT = `Analyze this food image and return ONLY valid JSON with exactly this structure:
{
    "foodName": "name of the food dish",
    "calories": 0,
    "protein": 0,
    "carbohydrates": 0,
    "fat": 0,
    "fiber": 0,
    "sugar": 0,
    "healthyScore": 0,
    "description": "Brief nutritional description",
    "portionSize": "Estimated portion size"
}
Only return the JSON object, no markdown, no other text.`;

    try {
        const imageData = fs.readFileSync(imagePath);
        const base64image = imageData.toString('base64');
        const mimeType = imagePath.endsWith('.png') ? 'image/png'
                       : imagePath.endsWith('.webp') ? 'image/webp' : 'image/jpeg';

        res.write(`event: status\ndata: ${JSON.stringify({ message: '🔍 Analyzing food with Groq Vision...' })}\n\n`);

        let resultData = null;
        let source = 'groq';

        if (process.env.GROQ_API_KEY) {
            try {
                const groqResult = await runGroqVisionAnalysis(FOOD_PROMPT, base64image, mimeType);
                if (groqResult?.success) {
                    resultData = groqResult.data;
                    source = groqResult.source;
                }
            } catch (groqErr) {
                console.warn('⚠️ Groq stream failed; falling back to Gemini:', groqErr.message);
            }
        }

        if (!resultData && process.env.GEMINI_API_KEY) {
            source = 'gemini';
            const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
            const model = genAI.getGenerativeModel({ model: GEMINI_MODEL });

            const result = await model.generateContent([
                FOOD_PROMPT,
                { inlineData: { mimeType, data: base64image } }
            ]);

            resultData = parseFoodAnalysisJson(result.response.text());
        }

        if (resultData) {
            const jsonStr = JSON.stringify(resultData, null, 2);

            // Stream the JSON in small chunks for UI effect
            const chunkSize = 10;
            for (let i = 0; i < jsonStr.length; i += chunkSize) {
                res.write(`event: chunk\ndata: ${JSON.stringify({ chunk: jsonStr.slice(i, i + chunkSize) })}\n\n`);
                await new Promise(r => setTimeout(r, 20));
            }

            res.write(`event: complete\ndata: ${JSON.stringify({ success: true, data: resultData, source })}\n\n`);
        } else {
            res.write(`event: error\ndata: ${JSON.stringify({ message: 'Could not parse food analysis response' })}\n\n`);
        }

    } catch (error) {
        console.error('❌ Food scan stream error:', error.message);
        const msg = error.message.includes('429')
            ? 'Rate limit exceeded. Please try again in a moment.'
            : error.message;
        res.write(`event: error\ndata: ${JSON.stringify({ message: msg })}\n\n`);
    }

    res.write(`event: done\ndata: ${JSON.stringify({ message: 'Analysis complete' })}\n\n`);
    res.end();
};

File: python-ai-services/app/services/ai_service.py
# app/services/ai_service.py
"""
AI service for chat (Groq) and food analysis (Gemini Vision)
"""
import os
import logging
import json
import time
import re
from typing import Optional, Generator, Dict, Any
from groq import Groq
from app.core.config import config

# ✅ Import Gemini with try/except
try:
    import google.generativeai as genai
    GENAI_AVAILABLE = True
except ImportError:
    GENAI_AVAILABLE = False
    logger = logging.getLogger(__name__)
    logger.warning("⚠️ google-generativeai not installed. Install with: pip install google-generativeai")

logger = logging.getLogger(__name__)

class AIService:
    """AI service for chat (Groq) and vision (Gemini)"""
    
    def __init__(self):
        # ============ Groq for Chat ============
        self.groq_api_key = os.getenv("GROQ_API_KEY", "")
        self.client = None
        self.is_loaded = False
        self.model = "llama-3.1-8b-instant"
        
        if self.groq_api_key and self.groq_api_key.startswith("gsk_"):
            try:
                self.client = Groq(api_key=self.groq_api_key)
                self.is_loaded = True
                logger.info("✅ Groq API initialized successfully!")
                logger.info(f"📦 Chat Model: {self.model}")
                self._test_connection()
            except Exception as e:
                logger.error(f"❌ Failed to initialize Groq: {e}")
                self.is_loaded = False
        else:
            logger.warning("⚠️ GROQ_API_KEY not set. Chat will use fallback.")
            logger.info("💡 Get your free API key from: https://console.groq.com")
        
        # ============ Gemini for Vision ============
        self.gemini_api_key = os.getenv("GEMINI_API_KEY", "")
        self.gemini_model = None
        self.gemini_available = False
        
        if GENAI_AVAILABLE and self.gemini_api_key:
            try:
                genai.configure(api_key=self.gemini_api_key)
                
                # ✅ Try multiple model names
                model_names = [
                    "gemini-1.5-pro",
                    "gemini-1.0-pro-vision",
                    "gemini-pro-vision",
                    "gemini-1.5-flash",
                ]
                
                for model_name in model_names:
                    try:
                        logger.info(f"🔄 Trying model: {model_name}")
                        self.gemini_model = genai.GenerativeModel(model_name)
                        # Quick test
                        test_response = self.gemini_model.generate_content("Hello")
                        if test_response:
                            self.gemini_available = True
                            logger.info(f"✅ Gemini API initialized with model: {model_name}")
                            break
                    except Exception as e:
                        logger.warning(f"⚠️ Model {model_name} not available: {e}")
                        continue
                
                if not self.gemini_available:
                    logger.error("❌ No Gemini vision model available")
                    logger.info("💡 Available models: https://ai.google.dev/gemini-api/docs/models")
                
            except Exception as e:
                logger.error(f"❌ Failed to initialize Gemini: {e}")
                self.gemini_available = False
        elif not GENAI_AVAILABLE:
            logger.warning("⚠️ google-generativeai package not installed.")
            logger.info("💡 Install with: pip install google-generativeai")
        else:
            logger.warning("⚠️ GEMINI_API_KEY not set. Food analysis will use fallback.")
            logger.info("💡 Get your free API key from: https://aistudio.google.com/apikey")
    
    def _test_connection(self):
        """Test if Groq API is working"""
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[{"role": "user", "content": "Hello"}],
                max_tokens=10,
            )
            logger.info("✅ Groq connection test successful!")
        except Exception as e:
            logger.error(f"❌ Groq connection test failed: {e}")
            self.is_loaded = False
    
    def load_model(self) -> bool:
        """No download needed - API ready"""
        return self.is_loaded
    
    def is_available(self) -> bool:
        """Check if chat service is available"""
        return self.is_loaded
    
    # =====================================================
    # ============ CHAT FUNCTIONS (Groq) ============
    # =====================================================
    
    def generate_response(self, prompt: str) -> Optional[str]:
        """Generate chat response using Groq API"""
        if not self.is_loaded or not self.client:
            logger.warning("Groq not available, using fallback")
            return None
        
        try:
            messages = [
                {
                    "role": "system",
                    "content": """You are NutriAI, a helpful health and nutrition assistant.
                    Provide clear, practical advice about health, fitness, and nutrition.
                    Keep responses concise, friendly, and actionable.
                    Never give medical advice - always recommend consulting a doctor for serious concerns."""
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ]
            
            response = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                temperature=0.7,
                max_tokens=300,
                top_p=0.9,
            )
            
            if response and response.choices:
                return response.choices[0].message.content.strip()
            
            return None
            
        except Exception as e:
            logger.error(f"❌ Groq error: {e}")
            return None
    
    def generate_response_stream(self, prompt: str) -> Generator[str, None, None]:
        """Generate streaming response using Groq API"""
        if not self.is_loaded or not self.client:
            logger.warning("Groq not available")
            yield "I apologize, but the AI service is currently unavailable. Please try again later."
            return
        
        try:
            messages = [
                {
                    "role": "system",
                    "content": """You are NutriAI, a helpful health and nutrition assistant.
                    Provide clear, practical advice about health, fitness, and nutrition.
                    Keep responses concise and actionable."""
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ]
            
            stream = self.client.chat.completions.create(
                model=self.model,
                messages=messages,
                temperature=0.7,
                max_tokens=300,
                top_p=0.9,
                stream=True,
            )
            
            for chunk in stream:
                if chunk.choices[0].delta.content:
                    yield chunk.choices[0].delta.content
                    
        except Exception as e:
            logger.error(f"❌ Streaming error: {e}")
            yield "I apologize, but I encountered an error. Please try again."
    
    # =========================================================
    # ============ FOOD ANALYSIS FUNCTIONS (Gemini) ============
    # =========================================================
    
    def analyze_food_image(self, image_base64: str, mime_type: str = "image/jpeg") -> Optional[Dict[str, Any]]:
        """
        Analyze food image using Gemini Vision with retry logic
        """
        if not self.gemini_available or not self.gemini_model:
            logger.warning("Gemini not available for vision")
            return None
        
        prompt = """Analyze this food image and return ONLY valid JSON with the following structure:
        {
            "foodName": "Name of the food dish",
            "calories": 0,
            "protein": 0,
            "carbohydrates": 0,
            "fat": 0,
            "fiber": 0,
            "sugar": 0,
            "healthyScore": 0,
            "description": "Brief description of the food and its nutritional value",
            "portionSize": "Estimated portion size"
        }
        
        Only return the JSON object, no other text."""
        
        max_retries = 3
        retry_delay = 2
        
        for attempt in range(max_retries):
            try:
                logger.info(f"🔄 Analyzing food image (attempt {attempt + 1}/{max_retries})...")
                
                image_part = {
                    "inline_data": {
                        "mime_type": mime_type,
                        "data": image_base64
                    }
                }
                
                response = self.gemini_model.generate_content([prompt, image_part])
                
                if response and response.text:
                    text = response.text
                    json_match = self._extract_json(text)
                    
                    if json_match:
                        data = json.loads(json_match)
                        logger.info(f"✅ Food analysis complete: {data.get('foodName', 'Unknown')}")
                        return data
                    else:
                        logger.warning("Could not extract JSON from response")
                
                return None
                
            except Exception as e:
                error_msg = str(e)
                
                if "429" in error_msg or "RESOURCE_EXHAUSTED" in error_msg:
                    logger.warning(f"⚠️ Rate limit hit (attempt {attempt + 1}/{max_retries})")
                    
                    if attempt < max_retries - 1:
                        wait_time = retry_delay * (2 ** attempt)
                        logger.info(f"⏳ Waiting {wait_time}s before retry...")
                        time.sleep(wait_time)
                        continue
                    else:
                        logger.error("❌ Max retries exceeded for rate limit")
                        return None
                else:
                    logger.error(f"❌ Gemini vision error: {error_msg}")
                    return None
        
        return None
    
    def analyze_food_image_stream(self, image_base64: str, mime_type: str = "image/jpeg") -> Generator[str, None, None]:
        """Stream food analysis response"""
        if not self.gemini_available or not self.gemini_model:
            yield f"event: error\ndata: {json.dumps({'message': 'Gemini not available'})}\n\n"
            return
        
        prompt = """Analyze this food image and return ONLY valid JSON with the following structure:
        {
            "foodName": "Name of the food dish",
            "calories": 0,
            "protein": 0,
            "carbohydrates": 0,
            "fat": 0,
            "fiber": 0,
            "sugar": 0,
            "healthyScore": 0,
            "description": "Brief description of the food and its nutritional value",
            "portionSize": "Estimated portion size"
        }
        
        Only return the JSON object, no other text."""
        
        try:
            image_part = {
                "inline_data": {
                    "mime_type": mime_type,
                    "data": image_base64
                }
            }
            
            yield f"event: status\ndata: {json.dumps({'message': '🔍 Analyzing food image...'})}\n\n"
            
            response = self.gemini_model.generate_content([prompt, image_part])
            
            if response and response.text:
                text = response.text
                json_match = self._extract_json(text)
                
                if json_match:
                    data = json.loads(json_match)
                    json_str = json.dumps(data, indent=2)
                    
                    chunk_size = 10
                    for i in range(0, len(json_str), chunk_size):
                        chunk = json_str[i:i+chunk_size]
                        yield f"event: chunk\ndata: {json.dumps({'chunk': chunk})}\n\n"
                        time.sleep(0.02)
                    
                    yield f"event: complete\ndata: {json.dumps({'success': True, 'data': data})}\n\n"
                else:
                    yield f"event: error\ndata: {json.dumps({'message': 'Could not parse response'})}\n\n"
            else:
                yield f"event: error\ndata: {json.dumps({'message': 'No response from Gemini'})}\n\n"
                
        except Exception as e:
            error_msg = str(e)
            if "429" in error_msg:
                yield f"event: error\ndata: {json.dumps({'message': 'Rate limit exceeded. Please try again in a moment.'})}\n\n"
            else:
                yield f"event: error\ndata: {json.dumps({'message': error_msg})}\n\n"
        
        yield f"event: done\ndata: {json.dumps({'message': 'Analysis complete'})}\n\n"
    
    def _extract_json(self, text: str) -> Optional[str]:
        """Extract JSON from text"""
        json_match = re.search(r'\{[\s\S]*\}', text)
        if json_match:
            return json_match.group(0)
        return None

# Singleton instance
ai_service = AIService()